### PR TITLE
Limit file activities to favorites

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Depends on https://github.com/owncloud/core/pull/15312
- [x] https://github.com/owncloud/activity/pull/271

### CODE REVIEW

File: apps/files/lib/activity.php [hunk 1]
```diff
@@ -270,9 +278,68 @@ public function filterNotificationTypes($types, $filter) {
 	 * @return array|false
 	 */
 	public function getQueryForFilter($filter) {
+		$user = \OC::$server->getActivityManager()->getCurrentUserId();
```
Comment (nickvergessen) [inline apps/files/lib/activity.php]: inject

File: apps/files/lib/activity.php [hunk 2]
```diff
@@ -270,9 +278,68 @@ public function filterNotificationTypes($types, $filter) {
...
 		}
+
+		if (!$user) {
+			// Remaining filters only work with a user/token
+			return false;
+		}
+
+		// Display actions from favorites only
+		if ($filter === self::FILTER_FAVORITES || $filter === 'all' && $this->userSettingFavoritesOnly($user)) {
+			$tagManager = \OC::$server->getTagManager();
```
Comment (nickvergessen) [inline apps/files/lib/activity.php]: inject

File: apps/files/lib/activity.php [hunk 3]
```diff
@@ -270,9 +278,68 @@ public function filterNotificationTypes($types, $filter) {
...
 	}
+
+	/**
+	 * Is the file actions favorite limitation enabled?
+	 *
+	 * @param string $user
+	 * @return bool
+	 */
+	protected function userSettingFavoritesOnly($user) {
+		return (bool) \OC::$server->getConfig()->getUserValue($user, 'activity', 'notify_stream_' . self::TYPE_FAVORITES, false);
```
Comment (nickvergessen) [inline apps/files/lib/activity.php]: inject

File: lib/private/activitymanager.php [hunk 4]
```diff
@@ -332,4 +358,43 @@ function getQueryForFilter($filter) {
...
+
+	/**
+	 * Get the user we need to use
+	 *
+	 * Either the user is logged in, or we try to get it from the token
+	 *
+	 * @return string
+	 * @throws \UnexpectedValueException If the token is invalid, does not exist or is not unique
+	 */
+	public function getCurrentUserId() {
```
Comment (nickvergessen) [inline lib/private/activitymanager.php]: tests have not been copied over

**Comment by karlitschek** [comment]
nice :+1:

File: lib/private/activitymanager.php [hunk 5]
```diff
@@ -12,8 +12,34 @@
...
+	protected $config;
+
+	/**
+	 * constructor of the controller
+	 *
+	 * @param IRequest $request
+	 * @param IUserSession $session
+	 * @param IConfig $config
+	 */
+	public function __construct(IRequest $request,
```
Comment (nickvergessen) [inline lib/private/activitymanager.php]: test needs to be adjusted

File: apps/files/lib/activity.php [hunk 6]
```diff
@@ -64,6 +84,7 @@ public function getNotificationTypes($languageCode) {
 		return [
 			self::TYPE_SHARE_CREATED => (string) $l->t('A new file or folder has been <strong>created</strong>'),
 			self::TYPE_SHARE_CHANGED => (string) $l->t('A file or folder has been <strong>changed</strong>'),
+			self::TYPE_FAVORITES => (string) $l->t('Limit notifications about creation and changes to your <strong>favorite files</strong>'),
```
Comment (nickvergessen) [inline apps/files/lib/activity.php]: Should mention that this only works for streams?

Comment (jancborchardt) [inline apps/files/lib/activity.php]: Alternatively just make it work for email as well. ;) I’d say if we put the »Limit to favorites« feature in the release, it needs to be for both Stream and Mail anyway.

Comment (nickvergessen) [inline apps/files/lib/activity.php]: have fun doing that in acceptable performance :P 
It's not like I'm not trying to, but I can't see the right way atm.

**Comment by nickvergessen** [comment]
Added tests, this should be ready now.

File: apps/files/lib/activity.php [hunk 7]
```diff
@@ -280,9 +306,64 @@ public function filterNotificationTypes($types, $filter) {
...
+			}
+			foreach ($favorites['folders'] as $favorite) {
+				$fileQueryList[] = '`file` LIKE ?';
+				$parameters[] = $favorite . '/%';
+			}
+
+			$parameters[] = 'files';
+
+			return [
+				' CASE WHEN `app` = ? THEN (' . implode(' OR ', $fileQueryList) . ') ELSE `app` <> ? END ',
```
Comment (nickvergessen) [inline apps/files/lib/activity.php]: The else needs to be dropped...

**Comment by karlitschek** [comment]
@nickvergessen who can help to review this?

**Comment by nickvergessen** [comment]
In order to test, you need to checkout `limit-file-activities-to-favorites` of the activity app aswell.

### How to test
0. Make sure `List your own actions in the stream` is enabled
1. Select the stream settings
2. Upload a file `SubFavo.txt` in a favorite folder (`Favorite/`)
3. Upload a file `SubNormal.txt` in a normal folder (`Normal/`)
4. Share the files as a link
5. Delete both files
6. Verify the displayed activities

### Test 1
1. Stream settings:
  - [x] A new file or folder has been created
  - [ ] Limit notifications about creation and changes to your favorite files (Stream only)
  - [x] A file or folder has been deleted
  - [x] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubNormal.txt and SubFavo.txt
  * You shared SubNormal.txt via link
  * You shared SubFavo.txt via link
  * You deleted SubNormal.txt and SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
  * You deleted SubNormal.txt and SubFavo.txt
4. Displayed activities in "Shares"
  * You shared SubNormal.txt via link
  * You shared SubFavo.txt via link

### Test 2
1. Stream settings:
  - [x] A new file or folder has been created
  - [x] Limit notifications about creation and changes to your favorite files (Stream only)
  - [x] A file or folder has been deleted
  - [x] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubFavo.txt
  * You shared SubNormal.txt via link
  * You shared SubFavo.txt via link
  * You deleted SubNormal.txt and SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
  * You deleted SubNormal.txt and SubFavo.txt
4. Displayed activities in "Shares"
  * You shared SubNormal.txt via link
  * You shared SubFavo.txt via link

### Test 3
1. Stream settings:
  - [x] A new file or folder has been created
  - [x] Limit notifications about creation and changes to your favorite files (Stream only)
  - [x] A file or folder has been deleted
  - [ ] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubFavo.txt
  * You deleted SubNormal.txt and SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
  * You deleted SubNormal.txt and SubFavo.txt
4. Displayed activities in "Shares"
  * *empty*


### Test 4
1. Stream settings:
  - [x] A new file or folder has been created
  - [x] Limit notifications about creation and changes to your favorite files (Stream only)
  - [ ] A file or folder has been deleted
  - [ ] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
4. Displayed activities in "Shares"
  * *empty*


### Test 5
1. Stream settings:
  - [x] A new file or folder has been created
  - [x] Limit notifications about creation and changes to your favorite files (Stream only)
  - [ ] A file or folder has been deleted
  - [ ] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
4. Displayed activities in "Shares"
  * *empty*


### Test 6
1. Stream settings:
  - [x] A new file or folder has been created
  - [ ] Limit notifications about creation and changes to your favorite files (Stream only)
  - [ ] A file or folder has been deleted
  - [ ] A file or folder has been shared
2. Displayed activities in "All" / "Activities by you"
  * You created SubNormal.txt and SubFavo.txt
3. Displayed activities in "Favorites"
  * You created SubFavo.txt
4. Displayed activities in "Shares"
  * *empty*


Feel free to imagine new cases when I forgot anything.

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by nickvergessen** [comment]
Fixed the new scrutinizer issues

**Comment by nickvergessen** [comment]
@LukasReschke second +1 ?

**Comment by MorrisJobke** [comment]
Reviewers needed

cc @Xenopathic @rullzer @LukasReschke @schiesbn

**Comment by MorrisJobke** [comment]
Ping :bell: 

One review left :)

**Comment by karlitschek** [comment]
here is mine :+1:

**Pull request merged**